gcp/gce_instance_test.go

```diff
@@ -0,0 +1,128 @@
+package gcp
+
+import (
+	"fmt"
+	"github.com/gruntwork-io/cloud-nuke/util"
+	gruntworkerrors "github.com/gruntwork-io/gruntwork-cli/errors"
+	"github.com/stretchr/testify/assert"
+	compute "google.golang.org/api/compute/v1"
+	"strings"
+	"testing"
+	"time"
+)
+
+func testZone() string {
+	return "us-central1-c"
+}
+
+func resourcesContains(resources []GcpResource, name string) bool {
+	for _, resource := range resources {
+		if resource.Name() == name && resource.Location() == testZone() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func createTestInstance(ctx *GcpContext, name string, protected bool) error {
+	machineType := fmt.Sprintf("projects/%s/zones/%s/machineTypes/f1-micro", ctx.Project, testZone())
+
+	iface := &compute.NetworkInterface{}
+
+	initializeParams := &compute.AttachedDiskInitializeParams{
+		SourceImage: "projects/debian-cloud/global/images/debian-9-stretch-v20181113",
+	}
+
+	disk := &compute.AttachedDisk{
+		AutoDelete:       true,
+		Boot:             true,
+		InitializeParams: initializeParams,
+	}
+
+	instance := &compute.Instance{
+		Name:               name,
+		MachineType:        machineType,
+		DeletionProtection: protected,
+		NetworkInterfaces:  []*compute.NetworkInterface{iface},
+		Disks:              []*compute.AttachedDisk{disk},
+	}
+
+	_, err := ctx.Service.Instances.Insert(ctx.Project, testZone(), instance).Do()
+	return err
+}
+
+func TestNukeInstances(t *testing.T) {
+	t.Parallel()
+
+	ctx, err := DefaultContext()
+	if err != nil {
+		assert.Fail(t, gruntworkerrors.WithStackTrace(err).Error())
+		return
+	}
+
+	instanceName := strings.ToLower("cloud-nuke-test-" + util.UniqueID())
+	protectedInstanceName := strings.ToLower("cloud-nuke-test-" + util.UniqueID())
+
+	defer func() {
+		call := ctx.Service.Instances.SetDeletionProtection(ctx.Project, testZone(), protectedInstanceName)
+		call.DeletionProtection(false)
+
+		_, err := call.Do()
+		if err != nil {
+			assert.Fail(t, gruntworkerrors.WithStackTrace(err).Error())
+		}
+
+		_, err = ctx.Service.Instances.Delete(ctx.Project, testZone(), protectedInstanceName).Do()
+		if err != nil {
+			assert.Fail(t, gruntworkerrors.WithStackTrace(err).Error())
+		}
```

Comment:
Extract the contents of this `defer` into a named function to make it clearer what it does.